circuit_knitting/cutting/cut_finding/cut_finding.py

```diff
@@ -0,0 +1,102 @@
+from .quantum_device_constraints import DeviceConstraints
+from .disjoint_subcircuits_state import DisjointSubcircuitsState
+from .circuit_interface import SimpleGateList
+from .lo_cuts_optimizer import LOCutsOptimizer
+from .cco_utils import qc_to_cco_circuit
+from ..instructions import CutWire
+from ..cutting_decomposition import cut_gates
+
+
+def find_cuts(
+    circuit: QuantumCircuit,
+    optimization: OptimizationSettings | dict[str, str | int],
+    constraints: DeviceConstraints | dict[str, int],
+) -> QuantumCircuit:
+    """
+    Find cut locations in a circuit, given optimization settings and QPU constraints.
+
+    Args:
+        circuit: The circuit to cut
+        optimization: Settings for controlling optimizer behavior. Currently,
+            only a best-first optimizer is supported. For a list of supported
+            optimization settings, see :class:`.OptimizationSettings`.
+        constraints: QPU constraints used to generate the cut location search space.
+            For information on how to specify QPU constraints, see :class:`.DeviceConstraints`.
+
+    Returns:
+        A circuit containing :class:`.BaseQPDGate` instances. The subcircuits
+        resulting from cutting these gates will be runnable on the devices
+        specified in ``constraints``.
+    """
+    circuit_cco = qc_to_cco_circuit(circuit)
+    interface = SimpleGateList(circuit_cco)
+
+    if isinstance(optimization, dict):
+        opt_settings = OptimizationSettings.from_dict(optimization)
+    else:
+        opt_settings = optimization
+
+    # Hard-code the optimization type to best-first
+    opt_settings.set_engine_selection("CutOptimization", "BestFirst")
+
+    if isinstance(constraints, dict):
+        constraint_settings = DeviceConstraints.from_dict(constraints)
+    else:
+        constraint_settings = constraints
+
+    # Hard-code the optimizer to an LO-only optimizer
+    optimizer = LOCutsOptimizer(interface, opt_settings, constraint_settings)
+
+    # Find cut locations
+    opt_out = optimizer.optimize()
+
+    wire_cut_actions = []
+    gate_ids = []
+
+    opt_out = cast(DisjointSubcircuitsState, opt_out)
+    opt_out.actions = cast(list, opt_out.actions)
+    for action in opt_out.actions:
+        if action[0].get_name() == "CutTwoQubitGate":
+            gate_ids.append(action[1][0])
+        else:
+            wire_cut_actions.append(action)
+
+    # First, replace all gates to cut with BaseQPDGate instances.
+    # This assumes each gate to cut is replaced 1-to-1 with a QPD gate.
+    # This may not hold in the future as we stop treating gate cuts individually.
+    circ_out = cut_gates(circuit, gate_ids)[0]
+
+    # Insert all the wire cuts
+    counter = 0
+    for action in sorted(wire_cut_actions, key=lambda a: a[1][0]):
+        if action[0].get_name() == "CutTwoQubitGate":
+            continue
+        inst_id = action[1][0]
+        qubit_id = action[2][0][0] - 1
+        circ_out.data.insert(
```

Comment:
I currently make no distinction between "left wire cut" and "right wire cut", which means I just place the cut wire on the index of the gate the wire cut references, which means it is always essentially a "left wire cut". I'm fairly sure this is an error. Noting here to follow up

Reply:
I'm less certain this is a bug after thinking about it, considering that I haven't encountered any problems separating the circuit under this logic. Any thoughts, @Ibrahim-Shehzad ?

Reply:
Perhaps I misunderstood your question but, in general, I don't think we can get away without specifying whether the first input wire to the gate is cut ("left wire cut") or the second input wire ("right wire cut"). I do think that it makes a difference. Was this what you were getting at?

Reply:
I think the "Left" and "Right" are confusing me because they aren't referencing which side of an operation to cut, they are referencing which qubit to cut with respect to some instruction. You always cut before the instruction, regardless of "Left" or "Right".

That seems to be what I implemented here, and it seems to always work, so I assume that is the logic. Does that sound right?

Reply:
> ...they aren't referencing which side of an operation to cut, they are referencing which qubit to cut with respect to some instruction. You always cut before (to the left of) the instruction, regardless of "Left" or "Right".

That is indeed correct.